### ISSUE TITLE
Keyboard focus/keyboard event inconsistancy

### ISSUE THREAD

**Issue by wmww** [MirServer/mir#2331]
#2245 we get
```
Attempt to send keyboard event to wl_surface@# even though it was not given keyboard focus
```
If Mir sends the Wayland frontend keyboard events for a surface that does not have keyboard focus. The problem is there is no synchronization between switching keyboard focus and sending input events, so events sent to the wrong surface could intermittently slip through (this has not been observed in the wild AFAIK). We should either restore the old behavior, or somehow ensure events arrive correctly.

**Comment by wmww** [comment]
This is happening when opening weston-terminal. Repro:
- From any terminal, run `weston-terminal` by pressing enter
- Occasionally, Mir will crash

**Comment by wmww** [comment]
Note that the #2330 may have introduced this in practice or made it more common since it rearranged timing and order of the relevant calls, but the underlying race condition was always present.

**Comment by AlanGriffiths** [comment]
> Fixed by #2339

I disagree. Reporting a problem quietly (as a log message) instead of loudly (fatal error) only reduces the severity of the manifestation.
